Switch perturbation from a passes algorithm to work with the standard engine

PaulTheLionHeart opened the pull request:
Outstanding issues:
Solid_guessing and boundary trace don't work in perturbation. Symmetry is turned off as it causes crashes.

PaulTheLionHeart committed:
Switch perturbation from a passes algorithm to work with the standard engine

Outstanding issues:
Solid_guessing and boundary trace don't work with perturbation.
Symmetry is turned off as it causes crashes.
Image tests are failing

LegalizeAdulthood commented:
Can you review the instructions here and make sure I have permission to make edits?

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

PaulTheLionHeart commented:
I have ticked the "Allow edits and access to secrets by maintainers.

LegalizeAdulthood commented:
Paul, I've rebased your branch onto master, resolving conflicts and editing the commit message to state the feature being added as opposed to the git mechanics.  I've also updated the title of the pull request to match.

LegalizeAdulthood commented:
I think the existing perturbation image tests are failing because they are still using `passes=p`.  Now would be a good time for you to get familiar with those image tests by looking at the CMakeLists.txt in tests/images/CMakeLists.txt.  I think we should switch the parameter `passes=p` to what we discussed with the parameter `perturbation=yes`

LegalizeAdulthood committed:
misc/Driver.h not misc/driver.h; sort includes in calcfrac.cpp

LegalizeAdulthood commented:
The build error on linux was due to incorrect case of header files.  Remember that the linux filesystem is case sensitive.

LegalizeAdulthood committed:
misc/Driver.h not misc/driver.h; sort includes in PertEngine.cpp

LegalizeAdulthood commented:
From what I can tell, here's a checklist of remaining work:
- [ ] Debug perturbation with solid guessing (`passes=g[123]`)
- [ ] Debug perturbation with boundary trace (`passes=b`)
- [ ] Update help documentation
- [ ] Remove `passes=p` as a valid option for passes, migrate any attempt to set `passes=p` to `perturbation=yes`
- [ ] Add `perturbation=yes/no/auto` parameter parsing
- [ ] When reading a saved GIF file, check for `passes=p` and migrate the setting to `g_perturbation`; this is only necessary if the `g_file_version` is Id 1.1

LegalizeAdulthood committed:
misc/Driver.h not misc/driver.h in solid_guess.cpp

PaulTheLionHeart commented:
I have changed the perturbation = AUTO, YES, NO and removed passes = p in my last pull request. Did it come through?

PaulTheLionHeart commented:
"If you update your branch locally from github, you should get my rebase.". When I checked your master, I didn't see the changes.  I'm a bit lost here.

LegalizeAdulthood commented:
I'm pushing the changes to your branch on your github fork.

You have to pull the changes to your local branch (Perturbation250209) from your fork on github.  Basically these commands:

```
git checkout Perturbation250209
git pull origin
```

... that is assuming that the remote named `origin` is connected to your github fork, which is the usual case.  Do `git remote -v` to see which remote names are attached to which repository.  For me, I have:

```
> git remote -v
origin  git@github.com:LegalizeAdulthood/iterated-dynamics (fetch)
origin  git@github.com:LegalizeAdulthood/iterated-dynamics (push)
paul    git@github.com:PaulTheLionHeart/iterated-dynamics (fetch)
paul    git@github.com:PaulTheLionHeart/iterated-dynamics (push)
```

which shows that for me, the remote named `origin` is associated with my repository on github and the remote named `paul` is associated with  your fork of my repository.

PaulTheLionHeart commented:
I was able to pull down the latest code once I knew t was branch Perturbation250209. 

Unfortunately, we have broken solid_guessing for normal operaton.
![fract001](https://github.com/user-attachments/assets/a6f2e35a-91b4-440c-87f1-d17ba13af813)
I copied the original master version of solid_guessing.cpp to this branch, rebuilt all and the error still exists, so it is not caused by anything I did in solid_guessing.cpp, but I suspect it's something in calcfrac.cpp. I'm out of my depth in understanding how this really works. I hope we can sort this out.

The cause of the crash when using symmetry in perturbation is line 1229 in calcfrac.cpp. This closes all the vectors before allowing the symmetry passes to work. I have commented it out. This doesn't fix symmetry, but stops a crash.

PaulTheLionHeart commented:
Ahhhh, on further exploration: pull down current master, the error shown in the attached GIF file exists in master. In fact, it exists in Id 1.1. Am I doing something silly or did thiis bug slip past testing? I tried Dosbox fractint and there was no issue.

LegalizeAdulthood commented:
OK, I started looking at the changes here and honestly.... the approach is all wrong.

The whole point of the standard engine is to generalize the escape-time fractal code so that the passes algorithms can be applied to the screen pixels according to the scanning algorithm defined by the passes algorithm.

Perturbation is just another way of computing an orbit, but you've got a whole bunch of conditional code bolted into various places where it shouldn't need to be.  The standard engine should just be invoking the orbit function and getting the resulting iteration count back out.  That's it.

I expected there to be some changes in the setup code, but these changes are too invasive and look too much like "perturbation owns the whole image" -- but it shouldn't.  Perturbation doesn't own the image and it doesn't control whether "pixels are complete", that's completely outside the responsibility of computing an orbit at a given location.

I don't like this term "glitched" as I've said before and I don't want to use it in Id.  It's non-specific jargon that doesn't reveal it's meaning or intent by reading it.  It isn't used at all in the original document describing the algorithm.

Whether it's less effort to continue editing this code, or just take the ideas and apply them from scratch to the standard engine is unclear.

I don't think these changes can be addressed in time to make it into 1.2, so it will have to wait until after that.

PaulTheLionHeart commented:
I agree with everything you said. I just can't find a way around it. If we didn't have to go back and redo the "errored" pixels things would be easier. I agree to defer perturbation until after 1.2. Do you have any intuition around how we can develop a better algorithm?

LegalizeAdulthood commented:
I'm doing a little work on getting kalles setup in Visual Studio (it uses a Makefile build) to see how it is computing reference points and so-on.

I'm also going to look at the Java source for SuperFractalThing to see how he's doing it.

PaulTheLionHeart commented:
I've been thinking about the high level design of perturbation...

I think we seen to take perturbation out of perform_work_list.

Here's what I mean...

perturbation()
{
    init_stuff();
    do
    {
        get_next_reference_coordinate();
        generate_reference_matrix();
        perform_work_list();
    }
    while (num_errir_pixels > total_pixels * PERCENTAGE_ALLOWED_ERRORS);
}

Does that make sense?

I still see some issues with marking pixels in blocks of pixels in guess() as okay or errored, but they shouldn't be too hard to fix once we get the high level algorithm sorted out.

LegalizeAdulthood commented:
BTW, now that 1.2 has been released, my suggestion is that you keep working on this branch.  All my wxWidgets 2.0 changes are taking place in the `wx` branch, so they shouldn't cause you any conflicts here.

PaulTheLionHeart commented:
Okay, I'll pull it down and set this as my new standard. What do you think of the high level way of dealing with perturbation?

LegalizeAdulthood commented:
> I've been thinking about the high level design of perturbation...
> 
> I think we seen to take perturbation out of perform_work_list.

I disagree; I don't think this is the right approach.  If you do this, then every passes algorithm has to be modified.

My gut tells me that the reason you want to put perturbation 'in charge' is because you want to manage the perturbation state at the outer loop and try to have perturbation drive the order in which pixels are evaluated.  This is the existing 'perturbation as passes' approach and it's what we're trying to get away from by integrating perturbation into the standard fractal engine.

Whatever perturbation state needs to be tracked should be tracked in `PerturbationEngine` as member variables that are reset when a new image is computed.  Basically, I'm suggesting an 'inversion of control' from the existing implementation.  The existing implementation drives the whole image computation by deciding which the order in which pixels are evaluated.  As pixels are evaluated, it incrementally chooses a new reference orbit when needed and tracks a _single_ reference orbit (and reference location) as a result.

To integrate perturbation into the standard fractal framewok, the perturbation code needs to be able to be called to evaluate any pixel in any order.  It will track a 'current reference orbit' (and location), but maintain multiple reference orbits as the needed.  When asked to compute a particular pixel, if that pixel is found to have too high an error for the existing reference orbit, then a new reference orbit needs to be chosen.  That could be one of the existing reference orbits that have already been created, or it could mean that a new reference orbit needs to be selected and added to the set of reference orbits seen so far.

Yes, it's a little more data in `PerturbationEngine` than the _single_ reference orbit tracked currently, but it makes perturbation independent of the order in which pixels are computed, which is the responsibility of the passes algorithms already implemented.

In addition, if you look more closely at the various passes algorithm, sometimes the order in which pixels are evaluated is implicit in the structure of the algorithm and not necessarily explicit as entries in the work list.  They may only be using the work list to support resuming a partially computed image.

PaulTheLionHeart commented:
Hi Richard,

 

My gut feeling is that this will require a greater number of reference pixels to be calculated and since this happens in arbitrary precision, van potentially increase the calculation time substantially.

 

There is also the issue of marking the block pixels as errored or not. This applies to whatever strategy we need.

 

Looking at how Kalles operates, they seem to have a kind of guessing algorithm that allows multi-threading, error management and even approximation. Unfortunately, these all run into one another when one looks at the source code and it is so poorly commented, it’s hard to know how the blocks fit together. The saving grace for Kalles is the blinding speed, especially at higher precision.

 

I must admit, I am still confused about the best implementation of perturbation. Rather than me going off on a tangent, we need to be clear about how we want to implement pert in Id. One thing I learned as a project manager was that 10% of effort spent in planning will affect 90% of the programming effort.

 

So let’s try to agree on the best algorithm.

 

Regards

 

Paul.

 

----------------------------------------------------------
Paul de Leeuw Computers     NSW Central Coast, Australia
Email:                             ***@***.*** ***@***.***> 
www:                           <https://deleeuw.com.au>
ABN                                         72 360 822 562
---------------------------------------------------------- 

 

 

From: Richard Thomson ***@***.***> 
Sent: Saturday, 15 March 2025 12:33 PM
To: LegalizeAdulthood/iterated-dynamics ***@***.***>
Cc: PaulTheLionHeart ***@***.***>; Author ***@***.***>
Subject: Re: [LegalizeAdulthood/iterated-dynamics] Switch perturbation from a passes algorithm to work with the standard engine (PR #256)

 

I've been thinking about the high level design of perturbation...

I think we seen to take perturbation out of perform_work_list.

I disagree; I don't think this is the right approach. If you do this, then every passes algorithm has to be modified.

My gut tells me that the reason you want to put perturbation 'in charge' is because you want to manage the perturbation state at the outer loop and try to have perturbation drive the order in which pixels are evaluated. This is the existing 'perturbation as passes' approach and it's what we're trying to get away from by integrating perturbation into the standard fractal engine.

Whatever perturbation state needs to be tracked should be tracked in PerturbationEngine as member variables that are reset when a new image is computed. Basically, I'm suggesting an 'inversion of control' from the existing implementation. The existing implementation drives the whole image computation by deciding which the order in which pixels are evaluated. As pixels are evaluated, it incrementally chooses a new reference orbit when needed and tracks a single reference orbit (and reference location) as a result.

To integrate perturbation into the standard fractal framewok, the perturbation code needs to be able to be called to evaluate any pixel in any order. It will track a 'current reference orbit' (and location), but maintain multiple reference orbits as the needed. When asked to compute a particular pixel, if that pixel is found to have too high an error for the existing reference orbit, then a new reference orbit needs to be chosen. That could be one of the existing reference orbits that have already been created, or it could mean that a new reference orbit needs to be selected and added to the set of reference orbits seen so far.

Yes, it's a little more data in PerturbationEngine than the single reference orbit tracked currently, but it makes perturbation independent of the order in which pixels are computed, which is the responsibility of the passes algorithms already implemented.

In addition, if you look more closely at the various passes algorithm, sometimes the order in which pixels are evaluated is implicit in the structure of the algorithm and not necessarily explicit as entries in the work list. They may only be using the work list to support resuming a partially computed image.

—
Reply to this email directly, view it on GitHub <https://github.com/LegalizeAdulthood/iterated-dynamics/pull/256#issuecomment-2726098505> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AJDBBKRUPIMC3TN5FRU2NYD2UN7NVAVCNFSM6AAAAABW3353SOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRWGA4TQNJQGU> .
You are receiving this because you authored the thread.  <https://github.com/notifications/beacon/AJDBBKRIWJUNHHPWLJQNTFT2UN7NVA5CNFSM6AAAAABW3353SOWGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTVCPT3ES.gif> Message ID: ***@***.*** ***@***.***> >

  <https://avatars.githubusercontent.com/u/4270863?s=20&v=4> LegalizeAdulthood left a comment (LegalizeAdulthood/iterated-dynamics#256) <https://github.com/LegalizeAdulthood/iterated-dynamics/pull/256#issuecomment-2726098505> 

I've been thinking about the high level design of perturbation...

I think we seen to take perturbation out of perform_work_list.

I disagree; I don't think this is the right approach. If you do this, then every passes algorithm has to be modified.

My gut tells me that the reason you want to put perturbation 'in charge' is because you want to manage the perturbation state at the outer loop and try to have perturbation drive the order in which pixels are evaluated. This is the existing 'perturbation as passes' approach and it's what we're trying to get away from by integrating perturbation into the standard fractal engine.

Whatever perturbation state needs to be tracked should be tracked in PerturbationEngine as member variables that are reset when a new image is computed. Basically, I'm suggesting an 'inversion of control' from the existing implementation. The existing implementation drives the whole image computation by deciding which the order in which pixels are evaluated. As pixels are evaluated, it incrementally chooses a new reference orbit when needed and tracks a single reference orbit (and reference location) as a result.

To integrate perturbation into the standard fractal framewok, the perturbation code needs to be able to be called to evaluate any pixel in any order. It will track a 'current reference orbit' (and location), but maintain multiple reference orbits as the needed. When asked to compute a particular pixel, if that pixel is found to have too high an error for the existing reference orbit, then a new reference orbit needs to be chosen. That could be one of the existing reference orbits that have already been created, or it could mean that a new reference orbit needs to be selected and added to the set of reference orbits seen so far.

Yes, it's a little more data in PerturbationEngine than the single reference orbit tracked currently, but it makes perturbation independent of the order in which pixels are computed, which is the responsibility of the passes algorithms already implemented.

In addition, if you look more closely at the various passes algorithm, sometimes the order in which pixels are evaluated is implicit in the structure of the algorithm and not necessarily explicit as entries in the work list. They may only be using the work list to support resuming a partially computed image.

—
Reply to this email directly, view it on GitHub <https://github.com/LegalizeAdulthood/iterated-dynamics/pull/256#issuecomment-2726098505> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AJDBBKRUPIMC3TN5FRU2NYD2UN7NVAVCNFSM6AAAAABW3353SOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRWGA4TQNJQGU> .
You are receiving this because you authored the thread.  <https://github.com/notifications/beacon/AJDBBKRIWJUNHHPWLJQNTFT2UN7NVA5CNFSM6AAAAABW3353SOWGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTVCPT3ES.gif> Message ID: ***@***.*** ***@***.***> >

LegalizeAdulthood commented:
> My gut feeling is that this will require a greater number of reference pixels to be calculated and since this happens in arbitrary precision, van potentially increase the calculation time substantially.

Maybe, but it's what the user asked for.  You get the existing behavior of computing pixels left to right, top to bottom by setting passes=1.  There are many things in Id that trade off computation time for accuracy.  Solid guessing can introduce errors from the 'exact' image.  Perturbation is just another approximation technique that trades off computation time for accuracy.

> There is also the issue of marking the block pixels as errored or not. This applies to whatever strategy we need.

This seems to be the key point in your approach that is in conflict with the idea of the passes in Id.

You keep coming back to wanting to 'mark pixels as errored', but I don't understand why this is needed.  What is the point?  Just handle the situation by adding a new reference point to the set as soon as a pixel is errored.  What are you going to do differently when you come back to this pixel that you aren't able to do the moment the pixel is discovered to be in error?

PaulTheLionHeart commented:
The central point seems to be how to find the best pixel for reference once we detect an error. The reason the block pixels were involved was because I don't want to look at every pixel to select the next potential reference pixel. 

"How a reference location is chosen for a pixel is arbitrary and up to the application" **_This_** is the key issue. Are you able to understand how SFT does it?

LegalizeAdulthood commented:
So if you had marked a region of pixels as errored out, what were you intending to do with that information?  Pick the 'center' of the region as the next reference location?

The SFT java code is pretty confusing; mostly because the responsibilities aren't isolated and the logic seems to be all over the place.  Without reverse engineering it completely, what I see is that the function `InitialiseCalculation` seems to start by choosing the reference point in the [center of the screen](https://github.com/LegalizeAdulthood/super-fractal-thing/blob/b35a9f82cb164a7a70380121ed0ad7b2169346a2/source/CalculationManager.java#L218), then [moving along the real axis](https://github.com/LegalizeAdulthood/super-fractal-thing/blob/b35a9f82cb164a7a70380121ed0ad7b2169346a2/source/CalculationManager.java#L235) and then finally, [spiraling out from the center point](https://github.com/LegalizeAdulthood/super-fractal-thing/blob/b35a9f82cb164a7a70380121ed0ad7b2169346a2/source/CalculationManager.java#L247).

You'd expect the class called `FindBestReferencePoint` to be the thing that actually finds the reference point, but as near as I can tell, that class just computes the orbit for the reference point, it doesn't actually do anything to find the 'best' reference point.

Naming things is important!

The algorithm I had in mind was:
1. Initialize by making the set of available reference points to be the point in the center of the image
2. If a pixel errors out, examine the set of existing reference points to find the one closest to the errored pixel.
   i. If that reference point is different from the current reference point, make it the current reference point and restart step 2
   ii. If that is the current reference point, then we need to add a new reference point; go to step 3.
3. Add new reference points by subdividing the screen into quadrants using the existing reference point as the center of a region.  So you start with a single reference point, then subdivide into 4 regions and add a new reference point at the center of the subdivided region containing the pixel and make it the current reference point.  Go back to step 2.

This implies that you need to track subdivided regions and their associated center points and which subdivided region has had it's center point computed as a reference orbit.

PaulTheLionHeart commented:
The process used by Pauldebrot in Fractal Forums is:

"The reference is escaping. Nanoscope's solution to this is to find blocks of pixels that have identical iteration counts down to the fractional part, which also catches most glitches. The largest connected component of the set of such pixels is found, then a pixel belonging to the block is found as near to its center of mass as possible, and then that pixel is calculated using normal arbitrary precision. That serves as the new reference (and also immediately does that one pixel). All other glitched/ref-escaped pixels are then recalculated using this reference. The test for such pixels is then repeated, and the whole procedure continues until they've all been corrected."

[https://fractalforums.org/index.php?topic=3849.msg24887#msg24887](url)

I have been trying to analyse Kalles code but get lost in the loss of comments and non-descriptive variable names.

LegalizeAdulthood commented:
I see nothing in that description that prevents you from implementing the same algorithm.  Note that he says:

> That serves as the new reference (and also immediately does that one pixel). **All other glitched/ref-escaped pixels are then recalculated using this reference.** 

So computing this 'block' of pixels at the same iteration count only serves to identify the new reference used for this pixel (and the pixels within the block).

This is all state internal to the `PertEngine` class.  When the scanning algorithm eventually gets around to asking for the orbits of pixels inside the connected component block associated with the reference point, you compute those pixels using that reference point.  If you want to do this approach, then you need to track irregularly shaped regions (the largest connected component) of pixel coordinates and their associated reference point.  The simplest way to do that is to have a vector of pixel coordinates that are sorted in such a way that you can binary search through the vector to determine if a pixel is in that list of points.

LegalizeAdulthood commented:
...in other words, there's no reason this algorithm can't be driven by the various passes options in Id.

LegalizeAdulthood commented:
In general, my suggestion is to implement the simplest thing that could possibly work.  **Don't worry about efficiency yet.**

First, strive for correctness, but implemented in the simplest possible way.

For instance, use a reference point until an orbit errors out.  Then just use the current pixel as the reference point and continue.

Maintain a single reference point that you use for pixels until they error out and then always replace the reference point with the current pixel.

This is extremely simple and will work and will still be faster than raw arbitrary precision for every pixel.

PaulTheLionHeart commented:
I actually tried the last idea and found that once I hit an errored section, I ended up needing a reference pixel every pixel until we reached the center of the region. It ended up being much slower than not using perturbation at all. The place where there are errors is a gradient and so each pixel as we go down the gradient will only be incrementally better than the previous one. I can code this if you like and we can see if there is a better algorithm later. Is it worth the effort?

LegalizeAdulthood commented:
If you compute a reference orbit for each pixel that should be _exactly_ the same as doing straight deep zooms.  How could it be **much** slower?  The perturbation overhead for such a simplistic implementation should be negligible.

PaulTheLionHeart commented:
How about I just code it and we can make adjustments from there. If it fixes some of the other issues with Guessing, it will be great. I'll start from master and only change what is needed. I'll let you know how I go.

LegalizeAdulthood commented:
I think that sounds like a good idea; let's get something working as simply as possible within the standard engine which will enable a bunch of features like periodicity checking inside/outside coloring, resuming, etc.

PaulTheLionHeart commented:
I'll do my best. I have 1 1/2 weeks of radiation left and will have a lot more time after that. I will try to get something to you as soon as I can.

PaulTheLionHeart commented:
My theory on why the calculation is a lot slower is that we calculate reference pixels right up to max_iteration whereas standard arbitrary precision pixels are only calculated until bailout.

LegalizeAdulthood commented:
> My theory on why the calculation is a lot slower is that we calculate reference pixels right up to max_iteration whereas standard arbitrary precision pixels are only calculated until bailout.

OK, that makes sense, but I think you can compute the reference orbit to bailout and then only compute more iterations on the reference orbit if the point being computed doesn't bailout before the reference orbit.

What is the likelihood of a point remaining within error limits **and** bailing out after the reference orbit?

PaulTheLionHeart commented:
This adds a lot of complication to the algorithm. For now, I will do the simplest thing I can to get perturbation working in the standard_calc method without worrying about speed. Then we can begin to optimize the reference point process.

LegalizeAdulthood commented:
> Please help me to find an algorithm for the pixel where an error first occurs.

I thought we decided that you would use the current pixel location as the new reference location, just to keep things as simple as possible.

LegalizeAdulthood commented:
What I was expecting was that if you needed to compute the new reference as the current pixel, then you'd compute the reference orbit and simply note which iteration escaped as you compute the orbit to max_iter.  Then the current pixel is colored with the iteration count that escaped.

PaulTheLionHeart commented:
I thought the same thing but somehow just can't make it work. Maybe I should wait until I feel better. I'm really very sick at the moment.

LegalizeAdulthood commented:
Take care of yourself first is what I say `:)`  This fractal stuff is fun, but unimportant in the big picture!

PaulTheLionHeart commented:
Thanks for understanding. I do love fractals and programming and it is a good distraction while I'm so sick. But I have to acknowledge my reduced thinking capacity right now. Today is my last radiation treatment and my body should begin to recover soon.

LegalizeAdulthood commented:
The first step in getting these pull requests to green builds is to get them rebased onto master.

PaulTheLionHeart commented:
There have been so many commits since these pull requests were made that it is probably best to close these ones and start again from pulling down the current master and adding changes from there. Then we can create a new pull request and rebase from there.

LegalizeAdulthood commented:
No worries, there's no rush on this.  I'm working on the wxwidgets stuff.

PaulTheLionHeart commented:
Thanks
